llvm/lib/Transforms/IPO/RepoPruning.cpp

```diff
@@ -150,11 +261,18 @@ static void addDependentFragments(
         NamedMDNode *const NMD = M.getOrInsertNamedMetadata("repo.definitions");
         assert(NMD && "NamedMDNode cannot be NULL!");
         NMD->addOperand(DMD);
+        // Collect the Dependent's XFixups.
+        auto DependentIt = Fragments->find(Repository, CM->digest);
+        assert(DependentIt != Fragments->end(Repository));
+        // Load  the dependent fragments if existing in the repository.
+        auto DependentFragment =
+            pstore::repo::fragment::load(Repository, DependentIt->second);
+        collectFragmentXFixupNames(Repository, DependentFragment, XFixupNames);
         // If function 'A' is dependent on function 'B' and 'B' is dependent on
         // function 'C', both RepoDefinition of 'B' and 'C' need to be added
         // into in the 'repo.definitions' during the pruning.
         addDependentFragments(M, DependentFragments, Fragments, Repository,
-                              CM->digest);
+                              XFixupNames, CM->digest);
```

Comment:
The comment immediately above this line of code is interesting. Why is this the case?

Reply:
The comment tries to explain why we need recursive calling `addDependentFragments` function. It gives an example to demo the case. By the way, it is not added by this PR. The test mentioned in the comment has been added into LLVM test when I fixed the dependent error.

Reply:
> The comment tries to explain why we need recursive calling addDependentFragments function. It gives an example to demo the case. By the way, it is not added by this PR.

Indeed but it does appear to very relevant indeed to the proposed change. In my plea above for help to understand what this does, this seemed particularly pertinent. For the definition of “dependent” in this code you appear to rely on the external fixups rather than the dependents section that I might expect.

Reply:
Sorry to misunderstand your question. Before discussing the `dependent`, I want to clarify the following things:
1)	Current Status: The discardable functions/variables are always pruned if they are present in the repository.
2)	The proposed change:  The discardable functions/variables are only pruned if they are present in the repository and referenced by other GOs which are in the repository.
3)	To check whether it is referenced by other pruned GOs. I use the external fixups.


Now, back to your question why I need to change `addDependentFragments` function to collect the external fixups of the fragments in the dependent section?

I want to use an example to explain the reason.

The initial build:

Assuming, 
1) there is a global function `A` and an internal function `B` during the RepoMetadataGeneration pass. 
2) a new function is created by other optimisation pass ( which is after the RepoMetadataGeneration pass) and added into the `A`’s dependent section. 
3) The output compilation contains three members: `A`, `B` and `D`. `B` is only referenced by `D` (not by `A`).
4) After the initial build, the `A`, `B` and `D` are saved in the repository. `A` does not have any external fixups.

Build the same file again. 

1) During the pruning stage, there are only `A` and `B` (`D` does not exist).
2) `A` is pruned and its external fixups are empty. 
3) `D` needs to be added into the RepoDefinition metadata and its external fixup needs to be collected. This is the change I added in the function of `addDependentFragments`: collect the external fixups of fragments in the dependent section. 
4) `B` is a discardable GO but it is referenced by `D`, therefore, it can be pruned.

Hope it makes sense. If not, maybe we can arrange a Team Meeting to discuss this in details. 😊